source/js/actions/index.js

```diff
@@ -3,15 +3,34 @@ import {
     EDIT_INPUT_ITEM,
     REMOVE_INPUT_ITEM,
     REQUEST_SAVED_STATE,
-    RECEIVE_SAVED_STATE
+    RECEIVE_SAVED_STATE,
+    SAVE_STATE,
+    SAVE_STATE_PENDING,
+    SAVE_STATE_COMPLETE
 } from '../constants/action-types';
 
-const formatSavedStateObject = savedStateObject =>
+const endpoint = '/api/state';
+
+const updateUrl = id => history.replaceState(null, null, id);
+
+const formatFirebaseObjectForState = savedStateObject =>
     Object.keys(savedStateObject).map(key => ({
         ...savedStateObject[key],
         id: key
     }));
 
+const formatStateArrayForFirebase = stateArray => {
+    let firebaseObject = {};
+    stateArray.forEach(stateObject => {
```

Comment:
This could be a reduce.

It doesn't have to be because it is readable as-is, but could be a candidate in the future.